### ISSUE TITLE
[Performance]: v0.23→v0.26: ~20% decode ITL regression on non-gated (relu2) BF16 MoE (SM100) — TRTLLM-gen backend swap (#43853) adds per-call host cost; --moe-backend flashinfer_cutlass recovers ~90%

### ISSUE THREAD

**Issue by dmvevents** [vllm-project/vllm#53472]
#### Summary

Between v0.23.0 and v0.26.0, decode inter-token latency on NVIDIA-Nemotron-3-Ultra-550B-A55B
(hybrid Mamba2 + attention, **non-gated** MoE with `mlp_hidden_act="relu2"`, BF16, TP8,
B200/SM100) regressed **+21%**, while prefill got **19% faster** and correctness was unaffected (greedy
outputs byte-identical on our benchmark gate prompt; one FP near-tie token flip on a second
prompt — detailed note in the Ablation section). We bisected the mechanism to the
unquantized-MoE backend selection:

- **#43853** (first in v0.24.0) flips `FlashInferTrtllmBf16Moe._supports_no_act_and_mul()`
  `False → True`. On v0.23.0 this predicate made TRTLLM **reject** non-gated models, so the
  oracle fell through to FlashInfer CUTLASS (priority #2). From v0.24.0 the TRTLLM-gen BF16
  monolithic kernel wins at priority #1 on SM100 for the same model.
- **#47427** (v0.26.0) additionally autotunes that kernel with
  `tune_max_num_tokens = max(max_num_tokens * dp_size, 8192)` — a bucket ≥8192 tokens — while
  a decode step at low concurrency runs at M≈1..10.

The swap runs on 48 of this model's 108 layers every decode step, which matches the observed
proportional (not fixed-offset) ITL regression.

**The new path is faster on the GPU and slower on the host.** Summed over the same 127-step decode
window, the MoE kernels the swap removed cost **605.0 ms** and the ones it added cost **326.9 ms**
(−46%). Over the same window the host-side wall time of `vllm::moe_forward_shared` rose from
**949 µs to 1102 µs median** (mean +175.7 µs) at an identical 48 calls/step — i.e. **+8.4–10.1 ms
per decode step, 55–63% of the measured +16.1 ms step delta, from this one op.** The extra host time
is *not* in vLLM's own wrapper: instrumented nested `cpu_op`s inside that op *fell* from 75 to 34 and
`cuda_runtime` share fell 13.2% → 5.2%, while **uninstrumented** time inside the op rose from
460 µs (48.4% of the call) to 846 µs (76.8%). The cost is inside an opaque FlashInfer frame.

That also explains the opposite-signed prefill result without invoking kernel shape: prefill
amortizes a fixed per-call host cost over ~1024 tokens and keeps the 46% GPU win, so it gets
**faster**; decode pays the same host cost per single token, so it gets **slower**.

#### Runtime evidence (selection lines, same model both versions)

v0.23.0:
```
INFO [unquantized.py:291] Using FlashInfer CUTLASS Unquantized MoE backend out of potential
backends: ['FlashInfer TRTLLM', 'FlashInfer CUTLASS', 'TRITON', 'BATCHED_TRITON'].
```
v0.26.0:
```
INFO [unquantized.py:374] Using MoEPrepareAndFinalizeNoDPEPMonolithic
INFO [unquantized.py:375] Using TrtLlmBf16ExpertsMonolithic MoE backend
```

#### Measured A/B (controlled: only the vLLM version moves; same node type, same engine args)

Serve args: `--tensor-parallel-size 8 --pipeline-parallel-size 1 --max-model-len 4096
--gpu-memory-utilization 0.97 --enforce-eager --mamba-cache-mode align` (enforce-eager ⇒
no cudagraphs/torch.compile in either leg). aiperf, ISL 1024 / OSL 512, concurrency 10,
n=100, 0 errors, warm (cold companion runs agree):

| metric | v0.23.0 | v0.26.0 | delta |
|---|---|---|---|
| ITL p50 (ms) | 89.40 | 107.97 | **+20.8%** |
| output tok/s | 110.79 | 90.15 | **−18.6%** |
| TTFT p50 (ms) | 503.52 | 407.99 | **−19.0% (faster)** |

The same delta reproduces on a 2-node disaggregated serving cell (ITL p50 49.55 → 61.25,
+23.6%) — the regression tracks the engine, not the serving topology. Greedy (temperature-0)
completions are byte-identical across versions.

#### torch.profiler A/B — per-kernel table

Same window on both legs: one 128-token temperature-0 completion from an 8-token prompt, warm,
`torch.profiler` on all 8 ranks, 127 decode steps each. Busiest-rank total CUDA kernel time
**15010.7 ms → 17261.2 ms (+15.0%)**.

**The MoE swap itself (GPU time went DOWN 46%):**

| | kernels | count each | total ms |
|---|---|---|---|
| v0.23 CUTLASS path (all GONE in 0.26) | 13 | 6144 | **605.0** |
| v0.26 TRT-LLM-Gen path (all NEW in 0.26) | 7 | 6096–6144 | **326.9** |

Removed in v0.26: 2× `cutlass::device_kernel<GemmUniversal<GroupProblemShape…>>` (150.60, 129.65 ms),
`vllm::moe::grouped_topk_fused_small_expert_count_kernel` (61.63),
`tensorrt_llm::…::blockExpertPrefixSumKernel<512>` (60.47),
`…::computeStridesTmaWarpSpecializedKernel` (51.16), `nvjet_sm100_tss_64x8_64x16_4x1_v_bz_splitK_TNT`
(32.54), `…::expandInputRowsKernel` (24.42), `…::mergeExpertPrefixSumKernel` (16.89),
`…::doActivationKernel` (15.36), `…::globalExpertPrefixSumKernel<512>` (13.91).

Added in v0.26: `bmm_Bfloat16_Bfloat16Bfloat16_Fp32_t128x8x128_s{4,5,6}…` (100.20, 85.08, 2.03,
1.94), `moe::dev::routing::routingCustom::routingIndicesBlockKernel` (64.36),
`moe::dev::finalize::finalizeKernel` (50.09), `cute_dsl_ll_bf16_dotprod` (23.21),
`vllm::activation_kernel<…relu_squared_kernel…>` (13.36).

**Where the +2250 ms actually lands — one row, and it is a spin-wait:**

| kernel | 0.23 count | 0.23 ms | 0.26 count | 0.26 ms | delta |
|---|---|---|---|---|---|
| `vllm::cross_device_reduce_1stage<__nv_bfloat16, 8>` | 13952 | 13635.48 | **13952** | 16076.25 | **+2440.77 (+17.9%)** |

Note the **call count is byte-identical** (13,952 = 109/step) — only the mean duration grew
(977.3 → 1152.3 µs). This is vLLM's custom one-shot TP all-reduce, whose kernel *busy-waits* for
peer ranks, so its duration is the accumulated straggler skew. It is the messenger, not the cause:
+19.07 ms/step of absorbed skew vs an independently measured **+18.57 ms/token** of aiperf ITL delta
(ratio 1.027 — the two agree to 2.7%). Every remaining matched-name row also rose at an identical
count (`splitKreduce_kernel` +39.5%, `_causal_conv1d_update_kernel` +11.0%, elementwise +12–49%),
which is the signature of a whole-step host-side stretch rather than any single kernel getting slower.

**Host-side counters over the same window (per decode step, busiest rank):**

| | v0.23.0 | v0.26.0 |
|---|---|---|
| host wall time per step | 114,752 µs | 130,867 µs (+14.0%) |
| kernel launches per step | 1549 | 1165 (−25%) |
| mean cost per launch API call | 5.47 µs | **9.29 µs (+70%)** |
| `vllm::moe_forward_shared` (48 calls/step) | 45,786 µs | 55,933 µs |
| `cudaEventSynchronize` | n=130, 659 µs | n=130, 667 µs |
| `Torch-Compiled Region` frames in trace | **18,560** | **128** |

v0.26 issues 25% *fewer* launches yet each launch API call costs 70% more. The last row is
unexpected and may be a second, separable issue — see Ask #3.

#### Ablation — `--moe-backend flashinfer_cutlass` on v0.26.0 recovers 86% of the step-time delta

Third profiled leg, same binary as the v0.26.0 leg, same window (2× 128-tok temp-0 warm, then one
128-token temp-0 completion under the profiler, 130 step annotations on all three legs). The flag
restores v0.23's exact selection — `Using MoEPrepareAndFinalizeNoDPEPModular` /
`Using FlashInferExperts MoE backend`.

| metric | v0.23.0 | v0.26.0 | v0.26.0 + flag | gap closed |
|---|---|---|---|---|
| step p50 (µs) | 114734.7 | 130840.8 | 117003.1 | **85.9%** |
| total CUDA kernel time (ms) | 15010.7 | 17238.6 | 15168.7 | 92.9% |
| `vllm::moe_forward_shared` p50 (µs) | 949.2 | 1101.6 | 971.9 | 85.1% |
| uninstrumented time inside that op (µs) | 466.1 | 876.3 | 494.2 | 93.1% |
| `cross_device_reduce_1stage` mean (µs) | 977.3 | 1151.7 | 989.9 | 92.8% |
| per-launch API cost (µs) | 5.47 | 9.29 | 5.80 | 91.4% |
| MoE GPU kernels (ms) | 229.9 (cutlass) | 325.4 (trtllm) | 233.8 (cutlass) | — |

Greedy output is byte-identical across all three legs. The residual ~14% is expected: this leg
still carries `enable_flashinfer_autotune=True` (so #47427's ≥8192-token tune bucket is still in
place) plus every non-MoE v0.23→v0.26 delta, and 23.0 ms of trtllm-marked kernels remain — the flag
redirects the fused-MoE experts call, not every TRT-LLM-derived kernel.

### The lost regional compilation is *coupled to the backend choice*, not a separate regression

| leg | `Torch-Compiled Region` frames | Dynamo cache lookups | MoE path |
|---|---|---|---|
| v0.23.0 | **18,560** | 18,560 | CUTLASS |
| v0.26.0 stock | **128** | 128 | TRT-LLM-Gen |
| v0.26.0 + `--moe-backend flashinfer_cutlass` | **18,560** | 18,560 | CUTLASS |

The per-region breakdown is byte-identical between v0.23.0 and the ablation leg
(`{0/1:1, 1/0:48, 2/0:48, 3/0:48, 0/2:127, 1/1:6096, 2/1:6096, 3/1:6096}`). So the regional
`torch.compile` frames are not independently dropped in v0.26 — they are absent **only** on the
TRT-LLM-Gen path and return in full when the CUTLASS backend is selected, on the same v0.26.0
binary with `--enforce-eager` set. The monolithic path bypasses the regionally-compiled MoE wrapper
altogether, which is the same underlying fact as the +70% per-launch API cost and the risen
uninstrumented in-op time: **one mechanism, not two.**

#### What we ruled out (evidence in hand, happy to share)

- Mamba2 SSD/conv kernels: byte-identical between tags (`mamba_mixer2.py`, `ops/ssd_*`,
  `causal_conv1d.py`; only a device-name-sanitization helper changed in `ops/mamba_ssm.py`).
- CUDA graph / compilation *policy*: both legs run `--enforce-eager` (CompilationMode.NONE,
  CUDAGraphMode.NONE). **Caveat:** this rules out the policy, not the presence of compiled code —
  see the `Torch-Compiled Region` row above and Ask #3.
- Model runner: NemotronH is not in `DEFAULT_V2_MODEL_RUNNER_ARCHITECTURES` at either tag —
  both use the V1 runner.
- NCCL / torch / triton wheels: identical (torch 2.11.0+cu130, triton 3.6.0, nccl 2.28.9).
- FlashInfer fused all-reduce default change: unreachable (`VLLM_ALLREDUCE_USE_FLASHINFER`
  defaults False at both tags).
- **The ReLU² custom-op change (#39058), live under eager:** its kernel is *present and cheap* in
  v0.26 (`vllm::activation_kernel<…relu_squared_kernel…>`, 13.36 ms / 6144 calls) and the
  elementwise kernels it replaced in v0.23 summed to more. Measured ~40% faster, not a regressor.
- **The `RoutedExperts` / `MoERunner` restructure adding host work:** refuted by host-op census —
  per MoE call, v0.26 executes *fewer* wrapper ops than v0.23 (`_maybe_dispatch` 13→11,
  `_apply_quant_method` 6→5, `_maybe_combine` 10→9). The new host cost is in the opaque FlashInfer
  frame, not in the restructured Python.
- **New per-step CPU sync points in the V1 `gpu_model_runner`:** `cudaEventSynchronize` is n=130 in
  *both* legs at ~660 µs total; the `torch.Event()` → `torch.cuda.Event(blocking=True)` change costs
  nothing measurable here.
- **`mamba2_attn` metadata build cost:** the change in this window *removes* device syncs
  (`async_tensor_h2d` replacing blocking `torch.tensor(...)`); `_compute_slot_mapping_kernel` went
  640 calls/1.85 ms → 128 calls/0.49 ms.

#### Ask

1. On SM100, should the non-gated (`RELU2_NO_MUL`) BF16 path consult an **M-dependent** heuristic
   rather than taking TRTLLM-gen unconditionally at priority #1? The GPU math is genuinely better,
   but at decode M≈1–10 the per-call host cost dominates and inverts the win. A shape-aware
   selection would keep the prefill gain and drop the decode loss.
2. Is the ~176 µs/call of host time in `TrtLlmBf16ExpertsMonolithic` expected? We cannot see inside
   the FlashInfer frame from a torch profile — if it is `AutoTuner.choose_one()` cache lookup or
   per-call workspace/stride setup, a cached fast path or a hoisted setup would recover the decode
   case without giving up the kernel. Note `#47427` tunes with
   `tune_max_num_tokens = max(max_num_tokens * dp_size, 8192)`, a bucket a low-concurrency decode
   never enters.
3. Is it intended that the monolithic path bypasses the regionally-compiled MoE wrapper? With
   `--enforce-eager` set, v0.23.0 (and v0.26.0 + the CUTLASS flag) execute 18,560
   `Torch-Compiled Region` frames nested *inside* `vllm::moe_forward_shared`; stock v0.26.0
   executes 128. The ablation shows this is **coupled to the backend selection**, not an
   independent change — so we are *not* filing it separately, but it looks like the concrete
   reason per-launch API cost rises 5.47 → 9.29 µs on 25% fewer launches, and it may be the
   cheapest lever: keeping the monolithic experts call inside the compiled region would plausibly
   recover the decode case without touching the kernel or the selection logic.
4. Workaround we're shipping meanwhile: `--moe-backend flashinfer_cutlass`, which recovers 86% of
   the step-time delta with byte-identical output (note `VLLM_USE_FLASHINFER_MOE_FP16` was removed,
   so the CLI flag is the only override).

### Your current environment (if you think it is necessary)

- vLLM v0.23.0 vs v0.26.0 (pip), flashinfer-python 0.6.12 / 0.6.14, torch 2.11.0+cu130,
  triton 3.6.0, CUDA 13.0, 8× B200 (SM100), single node for the agg cell.
- Model: nvidia/NVIDIA-Nemotron-3-Ultra-550B-A55B-BF16 (`n_routed_experts=512`, `top_k=22`,
  `moe_intermediate_size=5120`, `mlp_hidden_act="relu2"` → non-gated, 48 MoE / 48 Mamba2 /
  12 attention layers).
